clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java

```diff
@@ -263,13 +262,8 @@ public RecordsInfo info() {
         } else if (maxTimestamp == RecordBatch.NO_TIMESTAMP) {
             return new RecordsInfo(RecordBatch.NO_TIMESTAMP, lastOffset);
         } else {
-            long shallowOffsetOfMaxTimestamp;
-            // Use the last offset when dealing with record batches
-            if (compressionType != CompressionType.NONE || magic >= RecordBatch.MAGIC_VALUE_V2)
-                shallowOffsetOfMaxTimestamp = lastOffset;
-            else
-                shallowOffsetOfMaxTimestamp = offsetOfMaxTimestamp;
-            return new RecordsInfo(maxTimestamp, shallowOffsetOfMaxTimestamp);
+            // For create time, we always use offsetOfMaxTimestamp for the correct time -> offset mapping
+            return new RecordsInfo(maxTimestamp, offsetOfMaxTimestamp);
```

Comment:
> This has the same issue. The semantic for MAX_TIMESTAMP is the first offset with the max timestamp. So, if timestamp is TimestampType.LOG_APPEND_TIME, we need to use the baseOffset, instead of lastOffset.

Yes, updated.

> Also, could we remove the shallow part in RecordsInfo.shallowOffsetOfMaxTimestamp?
I've added a comment in another PR: https://github.com/apache/kafka/pull/15476#issuecomment-1990459827 . We can do all the renaming there.